File: Deepspace_Java/P2/src/deepspace/TestP2_.java
package deepspace;

import java.util.ArrayList;

/**
 * Class to test the second practice of the subject
 * @author Javier Gómez López
 */
public class TestP2_ {
    public static void main (String[] args){
        // Testing Hangar
        // ********************************
        System.out.println("Testing Hangar");
        System.out.println("****************************");
        
        //Creating a new hangar
        Hangar hangar_test = new Hangar(9);
        System.out.println("The created Hangar is:");
        System.out.println(hangar_test);
        
        //Testing the copy constructor
        Hangar hangar_copy = new Hangar(hangar_test);
        System.out.println("The copy of the previous Hangar is:");
        System.out.println(hangar_copy);
        System.out.println();
        
        //Testing UI version
        HangarToUI hangar_UI = hangar_test.getUIversion();
        System.out.println("The Hangar UI Version is; ");
        System.out.println(hangar_UI);
        System.out.println("-----------");
        System.out.println();
        
        
        //Testing add shield and weapon
        Weapon weapon_test = new Weapon("Weapon test", WeaponType.PLASMA, 2);
        ShieldBooster shield_test = new ShieldBooster("Shield test", 3.1415f,4);
        
        
        System.out.println("Adding some weapons and shields:");
        System.out.println();
        
        for(int i = 0; i < 5; i++){
            if(hangar_test.addShieldBooster(shield_test)){
                System.out.println("#" + i + " ShieldBooster added succesfully!");
            }
            else{
                System.out.println("#" + i + " ShieldBooster failed to add");
            }
        }
        
        System.out.println();
        
        //The last weapon has to fail because of the space available
        for(int i = 0; i < 5; i++){
            if(hangar_test.addWeapon(weapon_test)){
                System.out.println("#" + i + " Weapon added succesfully!");
            }
            else{
                System.out.println("#" + i + " Weapon failed to add");
            }
        }
        
        System.out.println();
        
        //Now show the copy
        System.out.println("State of the original hangar: ");
        System.out.println(hangar_test);
        System.out.println();
        
        System.out.println("State of the copy: ");
        System.out.println(hangar_copy);
        System.out.println("-----------");
        System.out.println();
        
        //Removing some weapons and shields
        System.out.println("Removing three shields...");
        for(int i = 0; i < 3; i++){
            hangar_test.removeShieldBooster(0);
        }
        
        System.out.println();
        
        System.out.println("Removing two weapons...");
        for(int i = 0; i < 2; i++){
            hangar_test.removeWeapon(0);
        }
        
        System.out.println();
        
        //Now show the copy
        System.out.println("State of the original hangar: ");
        System.out.println(hangar_test);
        System.out.println();
        
        System.out.println("State of the copy: ");
        System.out.println(hangar_copy);
        System.out.println("-----------");
        System.out.println();
        
        System.out.println();
        
        // Testing Damage
        // ********************************
        System.out.println("Testing Damage");
        System.out.println("****************************");
        
        ArrayList<WeaponType> weapon_type_list = new ArrayList<>(0);
        weapon_type_list.add(WeaponType.LASER);
        weapon_type_list.add(WeaponType.PLASMA);
        weapon_type_list.add(WeaponType.LASER);
        weapon_type_list.add(WeaponType.MISSILE);
        
        Damage damage_numeric = new Damage(4,3);
        Damage damage_specific = new Damage(weapon_type_list, 3);
        
        System.out.println("Numeric Damage: ");
        System.out.println(damage_numeric);
        System.out.println("Specific Damage: ");
        System.out.println(damage_specific);
        System.out.println("-----------");
        
        //Testing copy constructor
        Damage damage_numeric_copy = new Damage(damage_numeric);
        Damage damage_specific_copy = new Damage(damage_specific);
        
        System.out.println("The copy of the Numeric Damage is:");
        System.out.println(damage_numeric_copy);
        System.out.println("The copy of the Specific Damage is:");
        System.out.println(damage_specific_copy);
        System.out.println("-----------");
        System.out.println();
        
        //Testing UI Version
        DamageToUI damage_UI = damage_numeric.getUIversion();
        System.out.println("UI Numeric Version: ");
        System.out.println(damage_UI);
        System.out.println();
        
        //Testing adjust method
        System.out.println("Testing Damage.adjust()...");
        
        ArrayList<Weapon> weapons_to_adjust = new ArrayList<>(0);
        weapons_to_adjust.add(new Weapon("First weapon", WeaponType.LASER, 3));
        weapons_to_adjust.add(new Weapon("Second Weapon", WeaponType.PLASMA, 4));
        weapons_to_adjust.add(new Weapon("Third Weapon", WeaponType.MISSILE,1));
        
        ArrayList<ShieldBooster> shields_to_adjust = new ArrayList<>(0);
        shields_to_adjust.add(new ShieldBooster("First shield", 3.4f, 3));
        shields_to_adjust.add(new ShieldBooster("Second Shield", 6f, 1));
        
        Damage adjusted_damage = damage_numeric.adjust(weapons_to_adjust, shields_to_adjust);
        System.out.println("Numeric adjusted damage is ");
        System.out.println(adjusted_damage);
        System.out.println();
        
        adjusted_damage = damage_specific.adjust(weapons_to_adjust, shields_to_adjust);
        System.out.println("Adjusted specific damage is: ");
        System.out.println(adjusted_damage);
        System.out.println("-----------");
        System.out.println();
        
        
        //Testing dicard methods
        System.out.println("Discarding some weapons...");
        for(int i = 0; i < 2; i++){
            System.out.println("#" + i + " Discarding weapon on numeric damage");
            damage_numeric.discardWeapon(weapon_test);
            
            System.out.println("#" + i + " Discarding weapon on specific damage");
            damage_specific.discardWeapon(weapon_test);
        }
        
        System.out.println();
        
        System.out.println("Discarding some shieldboosters...");
        for(int i = 0; i < 2; i++){
            System.out.println("#" + i + " Discarding shield on numeric damage");
            damage_numeric.discardShieldBooster();
            
            System.out.println("#" + i + " Discarding shield on specific damage");
            damage_specific.discardShieldBooster();;
        }
        
        System.out.println();
        
        System.out.println("State of the numeric damage: ");
        System.out.println(damage_numeric);
        System.out.println("State of the copy of the numeric damage: ");
        System.out.println(damage_numeric_copy);
        
        System.out.println();
        
        System.out.println("State of the specific damage: ");
        System.out.println(damage_specific);
        System.out.println("State of the copy of the numeric damage: ");
        System.out.println(damage_specific_copy);
        System.out.println("-----------");
        System.out.println();
        
        System.out.println();
        
        // Testing EnemyStarShip
        // ********************************
        System.out.println("Testing EnemyStarShip");
        System.out.println("****************************");
        
        Damage damage_test = new Damage(2,3);
        Loot loot_test = new Loot(1,2,3,4,5);
        
        //Testing constructor
        EnemyStarShip enemy_test = new EnemyStarShip("Enemy Test", 1.5f, 3.14f, loot_test, damage_test);
        System.out.println("The testing EnemyStart is:");
        System.out.println(enemy_test);
        System.out.println();
        
        //Testing copy constructor
        EnemyStarShip enemy_copy = new EnemyStarShip(enemy_test);
        System.out.println("The copy of the enemy is: ");
        System.out.println(enemy_copy);
        System.out.println();
        
        //Testing the UI Version
        EnemyToUI enemy_UI = enemy_test.getUIversion();
        System.out.println("The UI version is: ");
        System.out.println(enemy_UI);
        System.out.println();
        
        //Testing some shots
        System.out.println("The enemy is going to receive some shots");
        for(int i = 0; i < 7; i++){
            ShotResult result = enemy_test.receiveShot(i);
            System.out.println("Receiving " + i + " power shot results in " + result);
        }
        
        System.out.println();
        
        // Testing SpaceStation
        // ********************************
        System.out.println("Testing SpaceStation");
        System.out.println("****************************");
        
        SuppliesPackage s = new SuppliesPackage(0.6f, 3.14f, 1.7f);
        SpaceStation space_test = new SpaceStation("Test Station", s);
        
        
        System.out.println("The SpaceStation is:");
        System.out.println(space_test);
        System.out.println("-----------");
        System.out.println();
        
        System.out.println("The speed of the SpaceShip is: " + space_test.getSpeed());
        System.out.println("-----------");
        System.out.println();
        
        //Testing validState
        System.out.println("With no pendingDamge, is the state valid?: " + space_test.validState());
        System.out.println("-----------");
        System.out.println();
        
        //Testing receive Hangar
        System.out.println("Testing receive Hangar");
        space_test.receiveHangar(hangar_test);
        System.out.println(space_test.getHangar());
        System.out.println("-----------");
        System.out.println();
        
        //Testing receiving Weapon and ShieldBoosters
        System.out.println("Adding some weapons and shieldboosters...");
        System.out.println("\nCleaning Hangar...");
        hangar_test = new Hangar(10);
        space_test.discardHangar();
        space_test.receiveHangar(hangar_test);
        
        space_test.receiveShieldBooster(shield_test);
        space_test.receiveWeapon(weapon_test);
        
        System.out.println(space_test.getHangar());
        System.out.println("-----------");
        System.out.println();
        
        //Testing mounting some Shields and Weapons
        System.out.println("Mounting some weapons and shieldboosters...");
        space_test.mountShieldBooster(0);
        space_test.mountWeapon(0);
        
        System.out.println("\nThe new Hangar:");
        System.out.println(space_test.getHangar());
        
        System.out.println("\nMounted Weapons and Shields:");
        System.out.println("\tWeapons:" + space_test.getWeapons());
        System.out.println("\tShields: " + space_test.getShieldBoosters());
        System.out.println("-----------");
        System.out.println();
        
        //Testing move
        System.out.println("Testing SpaceStation.move()");
        System.out.println("Fuel Units: " + space_test.getFuelUnits());
        System.out.println("Speed: " + space_test.getSpeed());
        space_test.move();
        System.out.println("Fuel units after movement: " + space_test.getFuelUnits());
        System.out.println("-----------");
        System.out.println();
        
        //Testing setPendingDamage
        System.out.println("Testing setPendingDamage");
        space_test.setPendingDamage(damage_test);
        System.out.println("\nNew damage:");
        System.out.println(space_test.getPendingDamage());
        
        System.out.println("Is the state Valid now? " + space_test.validState());
        System.out.println("-----------");
        System.out.println();
        
        //Testing discartes
        System.out.println("Restoring Hangar...");
        space_test.discardHangar();;
        space_test.receiveHangar(hangar_test);
        
        space_test.receiveShieldBooster(shield_test);
        space_test.receiveWeapon(weapon_test);
        
        System.out.println(space_test.getHangar());
        
        System.out.println("\nDiscarting some shields and weapons...");
        space_test.discardShieldBoosterInHangar(0);
        space_test.discardWeaponInHangar(0);
        
        System.out.println();
        System.out.println(space_test.getHangar());
        System.out.println("-----------");
        System.out.println();
        
        // Testing GameUniverse
        // ********************************
        System.out.println("Testing GameUniverse");
        System.out.println("****************************");
        
        GameUniverse universe_test = new GameUniverse();
        System.out.println("New GameUniverse: ");
        System.out.println(universe_test);        
        System.out.println("We cant test anymore because we can't init a game");
        System.out.println("-----------");
        System.out.println();
        
    }
}

File: Deepspace_Ruby/P2/TestP2_.rb
#!/usr/bin/env ruby
#encoding:utf-8

require_relative 'Hangar'
require_relative 'EnemyStarShip'
require_relative 'Damage'
require_relative 'GameUniverse'
require_relative 'SpaceStation'
require_relative 'DamageToUI'
require_relative 'EnemyToUI'
require_relative 'HangarToUI'
require_relative 'SpaceStationToUI'
<<<<<<< HEAD
=======
require_relative 'GameStateController'
require_relative 'SuppliesPackage'
>>>>>>> b56c07a8099700aa36fee02faffb2aadc4b580b0
require_relative 'Loot'

#Class for test main program for second practice
#
#@author Javier Gomez Lopez

class TestP2

<<<<<<< HEAD
=======
    include DeepSpace

>>>>>>> b56c07a8099700aa36fee02faffb2aadc4b580b0
    def self.main
        #Testing Hangar Class
        puts "TESTING Class: Hangar (3 times)"
        puts "**********************************"

        3.times do 
            maxElements = rand(10)
            puts "Creating Hangar.new(#{maxElements})"
            hangar_test = DeepSpace::Hangar.new(maxElements)
            puts "Result: "
            puts hangar_test
            
            puts

            puts "Adding seven different weapons"
            i = 0
            7.times do
                i += 1
                name = "Weapon#{i}"
                type = [DeepSpace::WeaponType::LASER, DeepSpace::WeaponType::MISSILE, DeepSpace::WeaponType::PLASMA][rand(3)]
                uses = rand(10)
                weapon = DeepSpace::Weapon.new(name,type,uses)
                puts "Trying to add: #{weapon}"

                if hangar_test.addWeapon(weapon) == false
                    puts "\t#{name} cannot be added on iteration #{i}"
                else
                    puts "\t#{name} can be added on iteration #{i}"
                end
            end

            puts

            puts "Hangar state is: #{hangar_test.to_s}"
            puts
            puts "Removing three weapons"
            3.times do
                removed_weapon = hangar_test.removeWeapon(0)
                puts "\tWeapon removed: #{removed_weapon}"
            end

            puts
            puts "Hangar state is:"
            puts hangar_test

            puts

            puts "Adding seven different shield boosters"
            i = 0
            7.times do
                i += 1
                name = "Shield#{i}"
                boost = rand(9) + rand()
                uses = rand(10)
                shield = DeepSpace::ShieldBooster.new(name, boost, uses)
                puts "Trying to add #{shield}"

                if hangar_test.addShieldBooster(shield) == false
                    puts "\t#{name} cannot be added on iteration #{i}"
                else
                    puts "\t#{name} can be added on iterarion #{i}"
                end
            end

            puts

            puts "Hangar state is:"
            puts hangar_test

            puts 

            puts "Removing two shields"
            2.times do
                removed_shield = hangar_test.removeShieldBooster(0)
                puts "\tShield removed: #{removed_shield}"
            end

            puts

            puts "Testing getters..."
            puts "\tWeapons: [#{hangar_test.weapons.join(", ")}]"
            puts "\tShields: [#{hangar_test.shieldBoosters.join(", ")}]"
            puts "\tMax elements: #{hangar_test.maxElements}"
            puts "\tUI Version: #{hangar_test.getUIversion}"
            puts
        end

        #Testing Damage Class
        puts "TESTING Class: Damage (3 times)"
        puts "**********************************"

        3.times do
            puts "Creating both types of damage..."
            nWeapons = rand(10)
            nShields = rand(10)

            puts

            puts "Creating Damage.newNumericWeapons(#{nWeapons}, #{nShields})"
            damage_numeric = DeepSpace::Damage.newNumericWeapons(nWeapons, nShields)
            weapons = []

            4.times do
                weapons << [DeepSpace::WeaponType::LASER, DeepSpace::WeaponType::MISSILE, DeepSpace::WeaponType::PLASMA][rand(3)]
            end

            nShields = rand(10)
            puts "Creating Damage.newSpecificWeapons([#{weapons.join(", ")}],#{nShields})"
<<<<<<< HEAD
                weapons << [DeepSpace::WeaponType::LASER, DeepSpace::WeaponType::MISSILE, DeepSpace::WeaponType::PLASMA][rand(3)]
                damage_specific = DeepSpace::Damage.newSpecificWeapons(weapons, nShields)
=======
            damage_specific = DeepSpace::Damage.newSpecificWeapons(weapons, nShields)
>>>>>>> b56c07a8099700aa36fee02faffb2aadc4b580b0

            puts

            puts "Numeric Damage:"
            puts "\t#{damage_numeric}"
            puts
            puts "Specific Damage:"
            puts "\t#{damage_specific}"
            puts 
            puts "UI version numeric: #{damage_numeric.getUIversion}"
            puts "UI version specific: #{damage_specific.getUIversion}"
            puts

            puts "Creating a copy of Numeric Damage"
            copy_test = DeepSpace::Damage.newCopy(damage_numeric)
            puts "Copied Damage: #{copy_test}"
            puts

            puts "Creating a copy of a Specific Damage"
            copy_test = DeepSpace::Damage.newCopy(damage_specific)
            puts "Copied Specific Damage: #{copy_test}"
            puts

            puts "Testing adjust method"

            adjust_weapons = []
            shields_adjust = []

            adjust_weapons << DeepSpace::Weapon.new("First Weapon", DeepSpace::WeaponType::LASER, 3)
            adjust_weapons << DeepSpace::Weapon.new("Second Weapon", DeepSpace::WeaponType::MISSILE, 5)
            adjust_weapons << DeepSpace::Weapon.new("Third Weapon", DeepSpace::WeaponType::PLASMA, 2)

            shields_adjust << DeepSpace::ShieldBooster.new("First Shield", 3.4, 3)
            shields_adjust << DeepSpace::ShieldBooster.new("Second shield", 6, 1)

            puts "damage_numeric.adjust([#{adjust_weapons.join(", ")}],[#{shields_adjust.join(", ")}])"
            puts
            puts "Damage before adjust:"
            puts damage_numeric
            puts "Damage after adjust:"
            puts damage_numeric.adjust(adjust_weapons,shields_adjust)
            puts

            puts "damage_specific.adjust([#{adjust_weapons.join(",")}],[#{shields_adjust.join(", ")}]"
            puts
            puts "Damage before adjust:"
            puts damage_specific
            puts "Damage after adjust:"
            puts damage_specific.adjust(adjust_weapons,shields_adjust)
            puts

            puts "Testing hasNoEffect"
            puts "Has no effect (numeric damage): #{damage_numeric.hasNoEffect}"
            noEffect = DeepSpace::Damage.newNumericWeapons(0,0)
            puts "Has no effect (empty damage): #{noEffect.hasNoEffect}"
            puts

            puts "Testing discardShieldBooster (10 shields)"
            10.times do
                damage_numeric.discardShieldBooster
            end

            puts "Damage Shields: #{damage_numeric.nShields}"

            puts

            puts "Testing discardWeapon (10 weapons)"
            weapon = DeepSpace::Weapon.new("Test Weapon", DeepSpace::WeaponType::PLASMA, 2)
            10.times do 
                damage_numeric.discardWeapon(nil)
                damage_specific.discardWeapon(weapon)
            end

            puts "Numeric Damage Weapons: #{damage_numeric.nWeapons}"
            puts "Specific Damage Weapons: [#{damage_specific.weapons.join(", ")}]"
            puts "--------------"
        end

        #Testing EnemyStarShip Class
        puts "TESTING Class: EnemyStarShip"
        puts "**********************************"

        puts "Creating an enemy starship"
        enemy_test = DeepSpace::EnemyStarShip.new("Enemy Test", 1.12, 3.14, DeepSpace::Loot.new(1,2,3,4,5), DeepSpace::Damage.newNumericWeapons(3,3))

        puts "Enemy starship: \n#{enemy_test}"
        puts

        puts "Crating a copy of the enemy starship"
        enemy_copy = DeepSpace::EnemyStarShip.newCopy(enemy_test)
        puts "Copy: \n#{enemy_copy}"
        puts

        puts "Creating UI Version..."
        enemy_UI = enemy_test.getUIversion
        puts "UI Version: \n#{enemy_UI}"
        puts

        puts "Protection: #{enemy_test.protection}"
        puts "Fire: #{enemy_test.fire}"
        puts

        puts "Testing EnemyStarShip.receiveShot()"
        10.times do |power|
            puts "\tAttacking with #{power} power!"
            result = enemy_test.receiveShot(power)
            puts "\tResult: #{result}"
        end
        puts "--------------"

        #Testing SpaceStation Class
        puts "TESTING Class: SpaceStation"
        puts "**********************************"

        s = DeepSpace::SuppliesPackage.new(0.6,3.14,1.7)
        space_test = DeepSpace::SpaceStation.new("Test Station", s)

        puts "The SpaceStation is:"
        puts space_test
        puts 

        puts "Testing Speed"
        puts "Speed: #{space_test.speed}"
        puts

        puts "With no pendingDamage, is the state valid?: #{space_test.validState}"
        puts

        puts "Testing receive Hangar"
        hangar_station = DeepSpace::Hangar.new(10)
        space_test.receiveHangar(hangar_station)
        puts space_test.hangar
        puts

        puts "Adding some weapons and shieldboosters..."
        puts "Cleaning Hangar..."
        space_test.discardHangar
        hangar_station = DeepSpace::Hangar.new(15)
        space_test.receiveHangar(hangar_station)

        shield_test = DeepSpace::ShieldBooster.new("Shield Test", 3.1415, 4)
        weapon_test = DeepSpace::Weapon.new("Weapon Test", DeepSpace::WeaponType::PLASMA, 2)

        puts space_test.inspect
        puts space_test.hangar.maxElements

        space_test.receiveShieldBooster(shield_test)
        space_test.receiveWeapon(weapon_test)

        puts

        puts "Mounting some weapons and shieldboosters..."
        space_test.mountShieldBooster(0)
        space_test.mountWeapon(0)

        puts "The new Hangar:"
        puts space_test.hangar
        puts

        puts "Mounted Weapons and Shields:"
        puts "\tWeapons: [#{space_test.weapons.join(", ")}]"
        puts "\tShields: [#{space_test.shieldBoosters.join(",")}]"
        puts

        puts "Testing SpaceStation.move()"
        puts "Fuel Units: #{space_test.fuelUnits}"
        puts "Speed: #{space_test.speed}"

        space_test.move

        puts "Fuel units after movement: #{space_test.fuelUnits}"
        puts

        puts "Testing setPendingDamage"
        damage_test = DeepSpace::Damage.newNumericWeapons(2,3)
        space_test.setPendingDamage(damage_test)
        puts "New damage:"
        puts space_test.pendingDamage

        puts "Is the state Valid now?: #{space_test.validState}"
        puts
        
        puts "Restoring Hangar..."
        space_test.discardHangar
        space_test.receiveHangar(hangar_station)

        space_test.receiveShieldBooster(shield_test)
        space_test.receiveWeapon(weapon_test)

        puts space_test.hangar

        puts "Discarting some shields and weapons..."
        space_test.discardShieldBoosterInHangar(0)
        space_test.discardWeaponInHangar(0)

        puts space_test.hangar
        puts


        #Testing GameUniverse Class
        puts "TESTING Class: GameUniverse"
        puts "**********************************"

        puts "We cant test this class because of the nil pointers"
    end #End Main

end #End class

TestP2.main